src/engine/client/dl_main.cpp

```diff
@@ -37,40 +37,207 @@ Maryland 20850 USA.
+	}
 
-	Cvar_SetValue( "cl_downloadCount", ( float ) dlnow );
-	return 0;
+	bool SetOptions(Str::StringRef url) {
+#define SETOPT(option, value) \
+if (curl_easy_setopt(request_, option, value) != CURLE_OK) { \
+	downloadLogger.Warn("Setting " #option " failed"); \
+	return false; \
 }
 
+		SETOPT( CURLOPT_USERAGENT, Str::Format( "%s %s", PRODUCT_NAME "/" PRODUCT_VERSION, curl_version() ).c_str() )
+		SETOPT( CURLOPT_REFERER, Str::Format("%s%s", URI_SCHEME, Cvar::GetValue("cl_currentServerIP")).c_str() )
+		SETOPT( CURLOPT_URL, url.c_str() )
+		SETOPT( CURLOPT_PROTOCOLS, long(CURLPROTO_HTTP) )
+		SETOPT( CURLOPT_WRITEFUNCTION, curl_write_callback(LibcurlWriteCallback) )
+		SETOPT( CURLOPT_WRITEDATA, static_cast<void*>(this) )
+		SETOPT( CURLOPT_FAILONERROR, 1L )
+		return true;
+	}
+};
+
+class FileDownload : public CurlDownload {
+	FS::File file_;
+
+	dlStatus_t WriteCallback(const char* data, size_t len) override {
+		try {
+			file_.Write(data, len);
+		} catch (std::system_error& e) {
+			downloadLogger.Notice("Error writing to download file: %s", e.what());
+			return dlStatus_t::DL_FAILED;
+		}
+		cl_downloadCount.Set(cl_downloadCount.Get() + len);
+		return dlStatus_t::DL_CONTINUE;
+	}
+
+public:
+	FileDownload(Str::StringRef url, FS::File file) : CurlDownload(url), file_(std::move(file)) {}
+};
+
+// If servers could ask the client to download any URL, there would be a security issue: the URL
+// could point to something on a private network that the server shouldn't have access to. Therefore,
+// before an HTTP download commences, the client checks for a magic file named PAKSERVER in the same
+// directory where the desired file is stored. If the file begins with the right magic string (we
+// check only a prefix to avoid any fiddly newline issues), this is interpreted as permission
+// Daemon clients to download anything from that directory.
+//
+// Note that this is the same issue addressed by the Same-Origin Policy in web browsers. In that
+// model, CORS headers are used to give scripts permission to access resources. We couldn't really
+// implement that here since there is no origin tracking of cgame binaries.
+const Str::StringRef PAKSERVER_FILE_NAME = "PAKSERVER";
+const Str::StringRef PAKSERVER_FILE_CONTENT_PREFIX = "This server is configured to host Daemon engine packages";
+class PakserverCheck : public CurlDownload {
+	Str::StringRef unmatchedPrefix_ = PAKSERVER_FILE_CONTENT_PREFIX;
+
+	dlStatus_t WriteCallback(const char* data, size_t len) override {
+		size_t n = std::min(len, unmatchedPrefix_.size());
+		if (0 != memcmp(unmatchedPrefix_.c_str(), data, n)) {
+			return dlStatus_t::DL_FAILED;
+		}
+		unmatchedPrefix_ = unmatchedPrefix_.suffix(n);
+		return unmatchedPrefix_.empty() ? dlStatus_t::DL_DONE : dlStatus_t::DL_CONTINUE;
+	}
+
+public:
+	PakserverCheck(Str::StringRef url) : CurlDownload(url) {}
+};
+
+struct DownloadState {
+	Util::optional<PakserverCheck> pakserverCheck;
+	Util::optional<FileDownload> realDownload;
```

Comment:
as opposed to a fake download?

Reply:
`PakserverCheck` is also a subclass of `CurlDownload`, so, I guess.

Reply:
`actualDownload`? this seems to less induce the opposite would be "fake" or "wrong"

Reply:
Done.